### ISSUE TITLE
Rewrite gosaml library

### ISSUE THREAD

**Issue by klizhentas** [gravitational/teleport#5171]
### Feature Request

We've had enough issues with go-saml library we are using. It is using GO XML library that is not 
[suited well](https://news.ycombinator.com/item?id=25422111) for SAML stuff.

We should consider rewriting this library from scratch with security in mind.

**Comment by awly** [comment]
Note: this implies also rewriting https://github.com/russellhaering/goxmldsig
We have 1 vuln in goxmldsig and 1 vuln in gosaml2

**Comment by klizhentas** [comment]
@russjones now, as a part of security hardening you are doing, does it make sense to adopt 

https://github.com/mattermost/xml-roundtrip-validator

as a protection step?

**Comment by russjones** [comment]
@klizhentas Yes, that was done as part of the mitigation.

https://github.com/gravitational/teleport/blob/master/vendor/github.com/russellhaering/gosaml2/decode_response.go#L31
https://github.com/gravitational/teleport/blob/master/vendor/github.com/russellhaering/gosaml2/decode_response.go#L376

**Comment by flyinghermit** [comment]
Yup, removing `gosaml2` dependency in favor of `crewjam/saml`. Though the `crewjam/saml` uses https://github.com/russellhaering/goxmldsig under the hood to sign and validate SAML assertions so we will still have an implicit dependency on the `goxmldsig` library.

The xml validator in the `cewjam/saml` also uses https://github.com/mattermost/xml-roundtrip-validator so that is a bonus.

So far, I have managed to test the `gosaml2` replacement with Okta as the SAML IdP and it looks promising. I need to test with our other officially supported auth connectors before making the full switch.
